cooked-validators/src/Cooked/Tx/Constraints.hs

```diff
@@ -55,51 +45,56 @@ class ToLedgerConstraint constraint where
   --  easily spend from multiple scripts at the same time.
   toLedgerConstraint :: constraint -> LedgerConstraint a
 
-instance ToLedgerConstraint MiscConstraint where
-  extractDatumStr (SpendsScript _ _ (_, Pl.ScriptChainIndexTxOut _ _ (Right datum) _)) =
-    M.singleton (Pl.datumHash . Pl.Datum $ Pl.toBuiltinData datum) (show datum)
+instance ToLedgerConstraint MintsConstraint where
   extractDatumStr _ = M.empty
 
-  toLedgerConstraint (SpendsScript v r (oref, o)) = (lkups, constr)
+  toLedgerConstraint (MintsConstraint Nothing pol tName amount) = (lkups, constr)
+    where
+      lkups = Pl.mintingPolicy pol
+      constr =
+        Pl.mustMintValue $
+          Pl.assetClassValue (Pl.assetClass (Pl.mpsSymbol . Pl.mintingPolicyHash $ pol) tName) amount
+  toLedgerConstraint (MintsConstraint (Just red) pol tName amount) = (lkups, constr)
     where
+      lkups = Pl.mintingPolicy pol
+      constr =
+        Pl.mustMintValueWithRedeemer (Pl.Redeemer (Pl.toBuiltinData red)) $
+          Pl.assetClassValue (Pl.assetClass (Pl.mpsSymbol . Pl.mintingPolicyHash $ pol) tName) amount
+
+instance ToLedgerConstraint InConstraint where
+  extractDatumStr (SpendsScript _ _ spOut) = case spOut ^? spOutDatum of
+    Just d -> M.singleton (Pl.datumHash . Pl.Datum $ Pl.toBuiltinData d) (show d)
+    Nothing -> M.empty
+  extractDatumStr SpendsPK {} = M.empty -- is this accurate? There might be a datum on PK outputs! TODO
```

Comment:
If I am not mistaken, this does not matter much on spend constraints because the datum you might spend, regardless whether it is from a script or pk, most certainly already has an entry in the pretty printing map from when the associated `PaysScript` or `PaysPKWithDatum` happened.